ZuluContent/Engines/Craft/Core/CraftItem.cs

```diff
@@ -766,64 +767,119 @@ private int CheckHueGrouping(Item a, Item b)
-            var craftSkill = Skills.First(sk => sk.SkillToMake == craftSystem.MainSkill);
+            var craftSkillRequired = GetCraftSkillRequired(from, typeRes, craftSystem);
+            var exceptionalDifficulty = craftSkillRequired;
+            var points = gainSkills ? GetCraftPoints(from, typeRes, craftSystem) : 0;
+            var exceptionalChance = GetExceptionalChance(from, craftSystem, ref exceptionalDifficulty);
+            var resource = CraftResources.GetFromType(typeRes);
+            var resQuality = CraftResources.GetQuality(resource);
+
+            quality = resQuality;
+
+            if (exceptionalChance > Utility.RandomDouble() &&
+                from.ShilCheckSkill(craftSystem.MainSkill, exceptionalDifficulty, 0))
+            {
+                mark = 2;
+
+                quality = (double) (int) (quality * GetQualityBonus(from)) / 100;
+            }
+
+            return from.ShilCheckSkill(craftSystem.MainSkill, craftSkillRequired, points);
+        }
+
+        public int GetQualityBonus(Mobile from)
+        {
+            var armsLoreValue = from.Skills[SkillName.ArmsLore].Value;
+            var multiplier = 5 + (int) (armsLoreValue / 10);
+
+            from.FireHook(h => h.OnQualityBonus(from, ref multiplier));
 
-            if (GetExceptionalChance(craftSystem, chance, from) > Utility.RandomDouble())
-                quality = 2;
+            multiplier += 100;
 
+            return multiplier;
+        }
+
+        public int GetCraftItemSkillRequired(Mobile from, CraftSystem craftSystem)
+        {
+            var craftSkill = Skills.FirstOrDefault(sk => sk.SkillToMake == craftSystem.MainSkill);
             var maxSkill = (int) (craftSkill?.MaxSkill ?? 0);
-            return from.ShilCheckSkill(craftSystem.MainSkill, maxSkill, maxSkill * 15);
+            return maxSkill;
+        }
+
+        public int GetCraftSkillRequired(Mobile from, Type typeRes, CraftSystem craftSystem)
+        {
+            int craftSkillRequired;
+            var itemSkillRequired = GetCraftItemSkillRequired(from, craftSystem);
+
+            if (craftSystem == DefBlacksmithy.CraftSystem)
+            {
+                var resource = CraftResources.GetFromType(typeRes);
+                craftSkillRequired = itemSkillRequired + (int) (CraftResources.GetCraftSkillRequired(resource) / 3);
+            }
+            else
+            {
+                craftSkillRequired = itemSkillRequired;
+            }
+
+            if (craftSkillRequired < 1)
+                craftSkillRequired = 1;
+            else if (craftSkillRequired > 140)
+                craftSkillRequired = 140;
+
+            return craftSkillRequired;
+        }
+
+        public int GetCraftPoints(Mobile from, Type typeRes, CraftSystem craftSystem)
+        {
+            int points;
+            var itemSkillRequired = GetCraftItemSkillRequired(from, craftSystem);
+
+            if (craftSystem == DefBlacksmithy.CraftSystem)
```

Comment:
Was thinking about just making these calculations be a function on the CraftSystem itself... what do you think?

Reply:
Yeah makes sense they live with the associated system